scrapy/core/downloader/handlers/__init__.py

```diff
@@ -20,25 +22,40 @@ def __init__(self, crawler):
             # component (extension, middleware, etc).
             if clspath is None:
                 continue
-            cls = load_object(clspath)
-            try:
-                dh = cls(crawler.settings)
-            except NotConfigured as ex:
-                self._notconfigured[scheme] = str(ex)
-            else:
-                self._handlers[scheme] = dh
+            self._schemes[scheme] = clspath
 
         crawler.signals.connect(self._close, signals.engine_stopped)
 
+    def _get_handler(self, scheme):
+        """Lazy-load the downloadhandler for a scheme
+        only on the first request for that scheme.
+        """
+        if scheme in self._handlers:
+            return self._handlers[scheme]
+        if scheme in self._notconfigured:
+            return None
+        if scheme not in self._schemes:
+            self._notconfigured[scheme] = \
+                    'no handler available for that scheme'
+            return None
+
+        dhcls = load_object(self._schemes[scheme])
```

Comment:
dang, I hadn't thought of the hard-failing part. So would you say this is a bad change in general? (Maybe then we should just put all S3 handling stuff into an optional/addon package instead?)

And I see how I should catch any exceptions on the load_object part and stuff it into notconfigured in that event. But I'm not sure I got the part on how to log them.